common/failure/failure.go

```diff
@@ -69,19 +71,28 @@ func NewTimeoutFailure(message string, timeoutType enumspb.TimeoutType) *failure
 }
 
 func Truncate(f *failurepb.Failure, maxSize int) *failurepb.Failure {
+	return TruncateWithDepth(f, maxSize, 20)
+}
+
+func TruncateWithDepth(f *failurepb.Failure, maxSize, maxDepth int) *failurepb.Failure {
 	if f == nil {
 		return nil
 	}
 
-	newFailure := &failurepb.Failure{
-		Source: f.Source,
+	// note that bytes are given to earlier calls first, so call in order of importance
+	trunc := func(s string) string {
+		s = util.TruncateUTF8(s, maxSize)
+		maxSize -= len(s)
+		return s
 	}
 
+	newFailure := &failurepb.Failure{}
+
 	// Keep failure info for ApplicationFailureInfo and for ServerFailureInfo to persist NonRetryable flag.
 	if f.GetApplicationFailureInfo() != nil {
 		newFailure.FailureInfo = &failurepb.Failure_ApplicationFailureInfo{ApplicationFailureInfo: &failurepb.ApplicationFailureInfo{
 			NonRetryable: f.GetApplicationFailureInfo().GetNonRetryable(),
-			Type:         f.GetApplicationFailureInfo().GetType(),
+			Type:         trunc(f.GetApplicationFailureInfo().GetType()),
 		}}
```

Comment:
shall we account for the entire size of FailureInfo?

Reply:
If the goal is really to get the length of the encoded proto <= maxSize, then yeah, we'd have to do more work. I don't think it's a big deal as long as it's a constant amount, and maxDepth ensures that it's a constant amount. But I don't really know the whole context around how this is used.

Reply:
I added some hacky accounting for proto overhead